pkg/kubelet/cm/devicemanager/device_plugin_stub.go

```diff
@@ -96,10 +102,17 @@ func (m *Stub) Start() error {
 	return nil
 }
 
-// Stop stops the gRPC server
+// Stop stops the gRPC server. Can be called without a prior Start
+// and more than once. Not safe to be called concurrently by different
+// goroutines!
 func (m *Stub) Stop() error {
+	if m.server == nil {
+		return nil
+	}
 	m.server.Stop()
-	close(m.stop)
+	<-m.wait
```

Comment:
we probably want to timeout on this

Reply:
But how long should the code wait, and what should it do when the timeout occurs? Proceeding without really freeing the socket would bring back the original problem.

WaitGroup.Wait() can't time out, which to me looks like the right choice: we have to ensure (and do) that the goroutine will complete, and we need to wait for that.

Reply:
What I usually do is create a function to timeout on the `Wait` and return an error.

```golang
func waitTimeout(wg *sync.WaitGroup, timeout time.Duration) error {
    c := make(chan struct{})
    go func() {
        defer close(c)
        wg.Wait()
    }()
    select {
    case <-c:
        return nil
    case <-time.After(time.Second):
        return fmt.Errorf("Device Plugin timeout while waiting for server to stop")
    }
}
```